src/Platforms/MySQL/CharsetMetadataProvider/ConnectionCharsetMetadataProvider.php

```diff
@@ -11,13 +11,28 @@
 /** @internal */
 final class ConnectionCharsetMetadataProvider implements CharsetMetadataProvider
 {
-    public function __construct(private readonly Connection $connection)
+    public function __construct(private readonly Connection $connection, private bool $useUtf8mb3)
     {
     }
 
+    public function normalizeCharset(string $charset): string
+    {
+        if ($this->useUtf8mb3 && $charset === 'utf8') {
+            return 'utf8mb3';
+        }
+
+        if (! $this->useUtf8mb3 && $charset === 'utf8mb3') {
+            return 'utf8';
+        }
+
+        return $charset;
+    }
```

Comment:
I don't think that `CharsetMetadataProvider` is the right abstraction for this logic. The purpose of `CharsetMetadataProvider` is to query a piece of configuration from the server in a cachable manner. A normalization method is something entirely different.

Reply:
The more I think about it: This normalization logic belongs into the platform classes. And if you put it there, you don't need that detection logic and you don't that `useUtf8mb3` flag.